### PR TITLE
Test create2 deploying code to same address

### PR DESCRIPTION
**Pull request by ewilz**

### CODE REVIEW

File: src/tests/test_utils.rs [hunk 1]
```diff
@@ -35,3 +35,7 @@ where
     let mut contract = EvmContract::default();
     test(&mut contract)
 }
+
+pub fn clear_attached_deposit() {
```
Comment (prestwich) [inline src/tests/test_utils.rs]: Rename function. It doesn't clear the attached deposit, it changes the environment. Maybe call it reset context :thinking:

File: src/tests/mod.rs [hunk 2]
```diff
@@ -163,6 +165,62 @@ fn test_view_call() {
     })
 }
 
+#[test]
+fn test_accurate_storage_on_selfdestruct() {
+    test_utils::run_test(100, |contract|  {
```
Comment (prestwich) [inline src/tests/mod.rs:170]: You should be able to pass `0` here to get no attached deposit. If you need more expressivity (i.e. to have different balance and deposit), you can update the `test_utils::run_test` function to accept more arguments

**Review by prestwich — CHANGES_REQUESTED** [review]
This looks great. Can we get another test that does the same thing in solidity? We're currently testing behavior across multiple txns, and it'd be good to test within a single tx as well

here's a sketch:

```
function doubleDeployTest(bytes32 _salt, bytes memory _contractBytecode) public {
        SelfDestruct other = SelfDestruct(deploy(_salt, _contractBytecode));
        other.store_uint(5);
        require(other.storedUint() == 5, "pre-destruction wrong uint");

        other.destruction();
        require(address(other).balance == 0);
        
        deploy(deploy(_salt, _contractBytecode));
        require(other.storedUint() == 0, "post-redeploy wrong uint");
    }
```

File: src/tests/mod.rs (hunk 2 above)
Comment (ewilz) [reply src/tests/mod.rs:170]: I need the attached deposit for line 171

File: src/tests/test_utils.rs (hunk 1 above)
Comment (ewilz) [reply src/tests/test_utils.rs]: sure. technically the only thing it changes about the environment so far is the attached_deposit (and account balance which doesn't seem to do be doing anything for the tests), but I see what you're saying

File: src/tests/mod.rs (hunk 2 above)
Comment (ewilz) [reply src/tests/mod.rs:170]: it doesn't seem to be the number of args that's the issue, more so for the context to change throughout the test if necessary. in that case, an update might be to accept a closure for individual transactions that will involve custom contexts (and just have the rust equivalent of a before_each block with default context for each test as a whole)

Comment (prestwich) [reply src/tests/mod.rs:170]: ah fair point. I guess the closure pattern is not ideal here, and really we should make a shortcut for changing the test env

File: src/tests/test_utils.rs (hunk 1 above)
Comment (ewilz) [reply src/tests/test_utils.rs]: fiixed

**Review by prestwich — APPROVED** [review]

**Pull request merged**